### PR TITLE
JBTM-3437 JMS 2.0 integration

### PR DESCRIPTION
**Pull request by mmusgrov**
https://issues.redhat.com/browse/JBTM-3437

MAIN CORE !TOMCAT !AS_TESTS !RTS !JACOCO !XTS !QA_JTA !QA_JTS_JACORB !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !BLACKTIE !PERF !LRA !NO_WIN !DB_TESTS !mysql !db2 !postgres !oracle

### CODE REVIEW

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionProxy.java [hunk 1]
```diff
@@ -72,6 +72,16 @@ public Session createSession(boolean transacted, int acknowledgeMode) throws JMS
         return xaConnection.createSession(transacted, acknowledgeMode);
     }
 
+    @Override
+    public Session createSession(int sessionMode) throws JMSException {
+        return xaConnection.createSession(sessionMode);
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionProxy.java:77]: It should check the JTA transaction is active and register the XAResource just like the above method
```
createSession(boolean trasacted, int acknowledgeMode);
```

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionProxy.java [hunk 2]
```diff
@@ -72,6 +72,16 @@ public Session createSession(boolean transacted, int acknowledgeMode) throws JMS
...
+    }
+
+    @Override
+    public Session createSession() throws JMSException {
+        return xaConnection.createSession();
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionProxy.java:82]: Same here.

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java [hunk 3]
```diff
@@ -85,4 +86,24 @@ public Connection createConnection(String userName, String password) throws JMSE
         return connection;
     }
 
+    @Override
+    public JMSContext createContext() {
+        return xaConnectionFactory.createXAContext();
+    }
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java:92]: I think we have to register the XAResource just like we do in ```ConnectionProxy```

```
XAJMSContext xaContext = xaConnectionFactory.createXAContext();
JMSContext context = new JMSContextProxy(xaContext, transactionHelper);
if (checkJTATransactionIsActive) {
     regiterXAResource(xaContext.getXAResource();
}
return context;
```

Also I think we need to provide ```JMSContextProxy``` to wrap ```JMSContext``` which delist the XAResource in ```close()``` method just like ```SessionProxy```.

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java [hunk 4]
```diff
@@ -85,4 +86,24 @@ public Connection createConnection(String userName, String password) throws JMSE
...
+
+    @Override
+    public JMSContext createContext(String userName, String password) {
+        return xaConnectionFactory.createXAContext(userName, password);
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java:96]: Same as above.

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java [hunk 5]
```diff
@@ -85,4 +86,24 @@ public Connection createConnection(String userName, String password) throws JMSE
...
+    }
+
+    @Override
+    public JMSContext createContext(String userName, String password, int sessionMode) {
+        return xaConnectionFactory.createXAContext(userName, password).createContext(sessionMode);
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java:101]: Same as above.

File: ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java [hunk 6]
```diff
@@ -85,4 +86,24 @@ public Connection createConnection(String userName, String password) throws JMSE
...
+    }
+
+    @Override
+    public JMSContext createContext(int sessionMode) {
+        return xaConnectionFactory.createXAContext().createContext(sessionMode);
```
Comment (zhfeng) [inline ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/ConnectionFactoryProxy.java:106]: Same as above.

**Review by zhfeng — CHANGES_REQUESTED** [review]

**Comment by zhfeng** [comment]
Thanks @mmusgrov for working on this enhancement and I think it could be better if we can add some unit tests for the JMS 2.0

**Comment by gytis** [comment]
I second @zhfeng here. I think that both session and context should be enlisted and delisted in the new methods. For the JMSContext we'll need a new proxy class to handle the delisting.

**Comment by mmusgrov** [comment]
I've put the PR on hold while I do the integration tests.

**Comment by zhfeng** [comment]
@mmusgrov is there any update ?

**Comment by mmusgrov** [comment]
@zhfeng (and t-ice and hhoffstaette) will you indicate what the impact of not having JMS 2.0 integration is for you? Once we know that then we can schedule some resources to do the work. Alternatively, since this is an open source community project we do accept pull requests from the community.

**Comment by mmusgrov** [comment]
closed in favour of #1979

**Pull request closed without merging**